[feat] 스터디 가입 요청, 수락, 거절 API 추가

crtEvent opened the pull request:
# ✨ Key changes
- [x] #25 
  

# 👋 To reviewers
- 서비스 레이어에 인터페이스를 추가했는데, 로직을 어디까지 위임해야 할지 고민이에요.
  - find만 하고 Facade에서 추가 검증 로직을 구현할지, 검증 자체를 validator에 모두 위임할지?
- 예외처리는 아직 결정되지 않아 임의로 작성했습니다.

## 추가된 API
### 스터디 가입 요청 API
- Method: `POST`
- URI: `/api/vi/studies/:studyId/join-requests`

### 스터디 가입 요청 수락 API
- Method: `PATCH`
- URI: `/api/v1/studies/:studyId/join-requests/:requestId/accept`

### 스터디 가입 요청 거절 API
- Method: `PATCH`
- URI: `/api/v1/studies/:studyId/join-requests/:requestId/reject`

crtEvent committed:
feat: JoinRequest 추가

crtEvent committed:
feat: JoinRequestRepository 추가

crtEvent committed:
feat: JoinRequest 생성 API 추가

crtEvent committed:
feat: JoinRequest 수락 API 추가

crtEvent committed:
feat: JoinRequest 거절 API 추가

crtEvent committed:
refactor: Service 레이어에 interface 적용 및 로직 분리

crtEvent committed:
fix: save 메서드에 누락된Transactional 어노테이션 추가

crtEvent committed:
chore: schema.sql에 join_request 테이블 추가

crtEvent committed:
[refactor] 로그인, 회원가입 관련 피드백 반영 (#20)

* refactor: `@CreationTimestamp` 대신 `@CreatedDate` 사용

- `@CreationTimestamp`은 hibernate 어노테이션, `@CreatedDate`은 spring data jpa 어노테이션 이다
- 이 프로젝트에선 spring data jpa을 메인으로 사용하므로 spring data jpa 어노테이션인 `@CreatedDate`를 사용하기로 함

* refactor: Member 생성자 접근제어자 public에서 private으로 변경

- 외부에서 호출할 일이 없으므로 기본 접근제어자를 private으로 한다

* refactor: Controller class 단위의 RequestMapping 어노테이션 제거

- 검색의 편의를 위해 메서드 단위에 API URL을 입력하도록 한다

* rename: Login -> SignIn으로 이름 변경

* rename: Converter 위치 infrastructure로 이동

* rename: SignInArgumentResolver 내부 변수명 더 명확하게 변경

* rename: OAuthProvider 관련 객체 패키지 auth_member로 이동

* rename: EnableConfigurationProperties 제거

* fix: 로그아웃 응답 상태 코드 204 No Content에서 200 OK로 변경

* rename: GitHubOAuthRequestException 패키지 위치 global에서 auth_member로 변경

* feat: Member 도메인에 TimeAuditingBaseEntity 적용

jinny-l commented on src/main/java/com/flytrap/venusplanner/api/join_request/domain/JoinRequest.java:
```diff
@@ -0,0 +1,57 @@
+
+    public static JoinRequest create(Long studyId, Long memberId) {
+        return null;
+    }
```
왜 return null 이에요?

crtEvent replied:
TDD 시도하느라 처음엔 완성되지 않은 엔티티를 만들었어요

jinny-l commented on src/main/java/com/flytrap/venusplanner/api/join_request/business/service/JoinRequestCurdFacadeService.java:
```diff
@@ -0,0 +1,47 @@
+
+        if (memberStudyRepository.existsByStudyIdAndMemberId(studyId, memberId)) {
+            throw new StudyAlreadyJoinedException("이미 스터디에 가입되어 있습니다.");
+        }
```
꼼꼼한 예외처리! 👍
이런 로직들이 Validator로 들어가면 좋을 것 같네요

jinny-l commented on src/main/java/com/flytrap/venusplanner/api/join_request/business/service/JoinRequestCurdFacadeService.java:
```diff
@@ -0,0 +1,47 @@
+                                throw new DuplicateJoinRequestException("이미 가입 요청된 상태입니다.");
+                            }
+                        }
+                );
```
existBy로 불린값 체크하면 안되나용??
Pageable로 결과값 갖고 오는 이유가 궁금합니당

crtEvent replied:
- 가장 최근 1개의 요청을 가져오는 쿼리를 짜다보니 Pageable을 반환하게 되어 있어서 이렇게 했습니다.
- 그런데 지금 보니 그냥 대기 상태의 요청이 있는지만 existBy로 확인하면 되네요.
- 개발하면서 처음 생각과 다르게 코드를 짜게 되었는데 잘 동작하다 보니 미처 생각하지 못했네요

jinny-l commented on src/main/java/com/flytrap/venusplanner/api/join_request/infrastructure/repository/JoinRequestRepository.java:
```diff
@@ -1,7 +1,21 @@
+    Page<JoinRequest> findMostRecentByMemberIdAndStudyId(
+            @Param("memberId") Long memberId,
+            @Param("studyId") Long studyId,
+            Pageable pageable);
```
alias 쓸 때 단어로 명시하는게 좀 더 읽기 편하다고 생각하는데
고려 한번 해봐주세요~

crtEvent replied:
네, 동의합니다. request로 수정하겠습니다.

jinny-l commented on src/main/java/com/flytrap/venusplanner/api/join_request/presentation/controller/JoinRequestController.java:
```diff
@@ -0,0 +1,33 @@
+            @PathVariable Long studyId,
+            @SignIn SessionMember sessionMember
+    ) {
+        JoinRequest joinRequest = joinRequestCrudService
+                .createJoinRequest(studyId, sessionMember.id());
```
객체가 꼭 필요한게 아니면 리턴타입을 id로 해도 되지 않을까 합니당

jinny-l commented on src/main/java/com/flytrap/venusplanner/api/join_request/business/service/JoinRequestCurdFacadeService.java:
```diff
@@ -44,4 +49,30 @@ public JoinRequest createJoinRequest(Long studyId, Long memberId) {
+        if (!joinRequest.isWaiting()) {
+            throw new JoinRequestAlreadyHandledException("이미 처리된 요청입니다.");
+        }
+        joinRequest.accept();
```
스트림 체이닝으로 깔끔하게 처리하면 좋을 것 같은데 방법이 안 떠오르네요..ㅎㅎ

crtEvent replied:
```java
@Transactional
    public void acceptJoinRequest(Long studyId, Long requestId, Long memberId) {
        studyValidator.validateStudyExists(studyId);
        memberStudyValidator.validateMemberCanAcceptJoinRequest(memberId, studyId);

        JoinRequest joinRequest = joinRequestValidator.findById(requestId);

        if (!joinRequest.validateStudyIdMatch(studyId)) {
            throw new StudyMismatchException("요청한 스터디 ID와 가입 요청의 스터디 ID가 일치하지 않습니다");
        }

        if (!joinRequest.isWaiting()) {
            throw new JoinRequestAlreadyHandledException("이미 처리된 요청입니다.");
        }

        joinRequest.accept();
    }
```
지금은 이렇게 되어 있어요

jinny-l commented on src/main/java/com/flytrap/venusplanner/api/join_request/infrastructure/repository/JoinRequestRepository.java:
```diff
@@ -15,7 +15,7 @@ public interface JoinRequestRepository extends JpaRepository<JoinRequest, Long>
-            @Param("studyId") Long studyId,
+            @Param("memberId") long memberId,
+            @Param("studyId") long studyId,
             Pageable pageable);
```
Long을 사용하면 무슨 문제가 발생하나요??
저도 옛날에 그런적이 있었던 것 같은데 기억이...

crtEvent replied:
테스트 해보니 아무 문제 없네요. 왜 바꿨는지 까먹었습니다. 죄송해요...

jinny-l commented on src/main/java/com/flytrap/venusplanner/api/join_request/business/service/JoinRequestUpdater.java:
```diff
@@ -0,0 +1,8 @@
+package com.flytrap.venusplanner.api.join_request.business.service;
+
+import com.flytrap.venusplanner.api.join_request.domain.JoinRequest;
+
+public interface JoinRequestUpdater {
```
Updater도 따로 분리하신 이유가 궁금합니당

crtEvent replied:
도메인 하나만 다루는 쿼리는 모두 Validator, Reader, Updater interface를 만들어서 실제 여러 도메인을 조합하는 Facade에서 사용하고자 했습니다. 논의해봤으면 좋겠어요.

Validator 따로 둔 이유가 Facade에 필요한 메서드만 노출하기 위함인데, 검증 이외의 단일 도메인 CRUD도 나눠서 인터페이스로 만들어야 같은 효과를 누릴 수 있을 것 같다고 생각했어요

crtEvent committed:
[feat] GlobalExceptionHanlder 추가 요구 사항 구현 (#32)

* refactor: `@CreationTimestamp` 대신 `@CreatedDate` 사용

- `@CreationTimestamp`은 hibernate 어노테이션, `@CreatedDate`은 spring data jpa 어노테이션 이다
- 이 프로젝트에선 spring data jpa을 메인으로 사용하므로 spring data jpa 어노테이션인 `@CreatedDate`를 사용하기로 함

* feat: CustomException 추가

* feat: GeneralExceptionType 추가

* feat: GlobalExceptionHandler 추가

* feat: MemberExceptionType 추가 및 적용

* feat: AuthMemberExceptionType 추가 및 적용

* feat: AUTHENTICATION_FAILURE_EXCEPTION 적용

* rename: 예외 메서드명 컨벤션 대문자에서 카멜케이스로 변경

crtEvent committed:
[core] CORS 설정 추가 (#40)

* chore: Cors 설정 추가

* chore: EnableJpaAuditing 설정 중복 제거

crtEvent committed:
[chore] submodule 업데이트 (#51)

crtEvent committed:
[fix] 로그인 시 회원 정보를 업데이트 하도록 수정 (#33)

* refactor: `@CreationTimestamp` 대신 `@CreatedDate` 사용

- `@CreationTimestamp`은 hibernate 어노테이션, `@CreatedDate`은 spring data jpa 어노테이션 이다
- 이 프로젝트에선 spring data jpa을 메인으로 사용하므로 spring data jpa 어노테이션인 `@CreatedDate`를 사용하기로 함

* rename: SignInArgumentResolver 내부 변수명 더 명확하게 변경

* feat: Member 도메인에 TimeAuditingBaseEntity 적용

* fix: 로그인 시 회원 정보를 업데이트 하도록 수정

crtEvent committed:
fix: JoinRequest 상태 검증 시 findBy 대신 existBy 사용

crtEvent committed:
fix: 예외 처리 추가

crtEvent commented:
## 이전 변경사항을 적용하기 위해 Rebase 했는데 이전 커밋 내용이 모두 딸려와서 close하고 다시 올리겠습니다.
